### ISSUE TITLE
No tab completion for the verdi commands

### ISSUE THREAD

**Issue by szoupanos** [aiidateam/aiida_core#498]
On my desktop, I see that the autocompletion doesn't work at all.

I ensured, by executing again the

> $ pip install --user -U setuptools pip wheel virtualenv

, that the venv was installed correctly.

I installed (again) aiida as it is shown at http://aiida-core.readthedocs.io/en/latest/installation/index.html
> (aiidapy) $ pip install -e aiida_core

And I also executed a 

> verdi setup

 to reconfigure a specific profile (I don't know if this changes anything).

So after doing the above, I still don't have autocompletion when writing half of the command and I press tab. On my laptop, on the other hand, autocompletion works and it is actually fast.

I tested the above with 

> verdi profile

 & 

> verdi calculation

Any ideas?

Moreover, when I try to install AiiDA using a new (& different) virtual environment, installation fails.

Thanks a lot!
Spyros

**Comment by DropD** [comment]
@szoupanos Can I close this? I think it was solved, right?

**Comment by szoupanos** [comment]
Hello Rico,

no, I still have this problem.

**Comment by DropD** [comment]
Ok, unfortunately i have no time anymore to look into this, can you reassign please?

**Comment by fernandogargiulo1986** [comment]
In my laptop, I can reproduce the problem if I comment out this portion of code of ~/.bashrc, which is a legacy of a previous installation

```bash
###########################TAB COMPLETION IN VERDI COMMAND
function _aiida_verdi_completion
{
    OUTPUT=$( $1 completion "$COMP_CWORD" "${COMP_WORDS[@]}" ; echo 'x')
    OUTPUT=${OUTPUT%x}
    if [ -z "$OUTPUT" ]
    then
    # Only newline is a valid separator
        local IFS=$'\n'

        COMPREPLY=( $(compgen -o default -- "${COMP_WORDS[COMP_CWORD]}" ) )
    # Add either a slash or a space, depending on whether it is a folder
    # or a file. printf %q escapes the filename if there are spaces.
        for ((i=0; i < ${#COMPREPLY[@]}; i++)); do
            [ -d "${COMPREPLY[$i]}" ] && \
               COMPREPLY[$i]=$(printf %q%s "${COMPREPLY[$i]}" "/") || \
               COMPREPLY[$i]=$(printf %q%s "${COMPREPLY[$i]}" " ")
        done

    else
        COMPREPLY=( $(compgen -W "$OUTPUT" -- "${COMP_WORDS[COMP_CWORD]}" ) )
        # Always add a space after each command
        for ((i=0; i < ${#COMPREPLY[@]}; i++)); do
            COMPREPLY[$i]="${COMPREPLY[$i]} "
        done
    fi
}
complete -o nospace -F _aiida_verdi_completion verdi
```
otherwise it works fine.

**Comment by szoupanos** [comment]
I resolved the "problem". I suppose that I didn't read the documentation throughly or when I did, this part was not very clear.

More specifically, I used the 
eval "$(verdi completioncommand)"
line in my .bashrc to enable autocompletion.

This was sufficient until I switched to an AiiDA version with virtual environment.
After doing this transition (having AiiDA working in an VE), my .bashrc was complaining about a verdi command that was not found (the verdi completioncommand that was called during login) but I didn't invest time to clean the bashrc (or to activate the virtual environment before the evaluation of the verdi completioncommand.

To summarise if the 
eval "$(verdi completioncommand)"
is used, it is better to be added in the activation script of the virtual environment.

If the result of the verdi completioncommand is used, it can be added in both files (.bashrc & activation script of the virtual environment)
